### PR TITLE
Sensor Autonomous

### PR DESCRIPTION
**Pull request by BD103**
- [x] Sketch out an autonomous based on sensors that mostly works
- [x] Wrap all sensor loops in a maximum time loop so it is less likely to continue running into a wall
- [ ] ~90% accuracy on full power battery
- [x] Fix weird sensor bug that sometimes happen (it says `sensorBl` is broken or something)

### CODE REVIEW

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java [hunk 1]
```diff
@@ -37,13 +42,22 @@
...
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import lombok.Getter;
+
+public class Robot extends AbstractRobot implements WheeledRobot, StepWheeledRobot, ArmRobot, DistanceSensorRobot {
+    public static ExecutorService executorService;
+
     // Discuss if private is better idea
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:54]: if you're going to keep them public, it's a better idea to have them static. Else, use lombok getters and setters

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/PositionTelemetry.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import org.firstinspires.ftc.teamcode.api.config.Naming;
+import org.firstinspires.ftc.teamcode.api.hw.Gyroscope;
+
+@Config
+@TeleOp(name="Position Telemetry", group=Naming.OPMODE_GROUP_UTIL)
+public class PositionTelemetry extends LinearOpMode {
+    @Override
+    public void runOpMode() throws InterruptedException {
+        telemetry = new MultipleTelemetry(telemetry, FtcDashboard.getInstance().getTelemetry());
+        Robot.gyro1 = new Gyroscope(hardwareMap.get(BNO055IMU.class, "imu 1"), "gyro1");
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/PositionTelemetry.java]: can we use the naming file for this?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/hw/SmartServo.java [hunk 3]
```diff
@@ -99,25 +99,24 @@ private void scanServo(double position, int ms) {
...
-                scanServo(position, ms);
-            }
-        });
-        lock = false;
+        locked = true;
+        Robot.executorService.execute(() -> scanServo(position, ms));
+        locked = false;
     }
 
     public void scanServoSync(double position, int ms) {
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/hw/SmartServo.java]: Let's add another function that "steps" the servo with a time interval (ie step the servo, note down the last used time. If the last used time is far enough, then step again. Increasing the time interval can make movements slower, and thus can be controlled from teleop in a semi-async method in that it doesn't block the main thread but we can enforce sleeps on it.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/autonomous/TimedSmartDuck.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+        robot.armLeft.setPwmEnable();
+        robot.armRight.setPwmEnable();
+        robot.clawLeft.setPwmEnable();
+        robot.clawRight.setPwmEnable();
+
+        // Move left a little bit
+        ElapsedTime runtime = new ElapsedTime();
+
+        while (runtime.seconds() < 0.25 && opModeIsActive()) {
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/autonomous/TimedSmartDuck.java]: For this whole file, you don't need the while loops. Just set the power to 0.5, sleep, then set the power to 0.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/ColorSensorTelemetry.java [hunk 5]
```diff
@@ -61,4 +68,13 @@ public void runOpMode() throws InterruptedException {
             telemetry.update();
         }
     }
+    static class Sensor {
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/ColorSensorTelemetry.java]: Why is this here?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/Control.java [hunk 6]
```diff
@@ -0,0 +1,4 @@
+package org.firstinspires.ftc.teamcode.opmode.teleop;
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/Control.java]: What is this file going to do?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpAlt.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
+package org.firstinspires.ftc.teamcode.opmode.teleop;
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpAlt.java]: I like what you are doing with the event-driven stuff. It does bring up the question of how we are going to implement something if we need to run a loop (ie position tracking, etc) which won't be event-driven.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpMain.java [hunk 8]
```diff
@@ -67,11 +62,11 @@ public void runOpMode() {
             }
 
             if (gamepad1.dpad_up) {
-                robot.armLeft.scanServoAsync(robot.armLeft.getPosition() + 0.01, 20);
-                robot.armRight.scanServoAsync(robot.armRight.getPosition() + 0.01, 20);
+                robot.armLeft.setPosition(Range.clip(robot.armLeft.getPosition() - 0.001, 0.6, 1));
```
Comment (9p4) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpMain.java]: Look at the feedback on the SmartServo class. TL;DR: we need to make a non-blocking function that checks if enough time has elapsed so we can control the servos better.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java (hunk 1 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:54]: I have no idea what that's used for, so...

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/PositionTelemetry.java (hunk 2 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/PositionTelemetry.java]: That is out of the scope of the PR, maybe do a different commit?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/autonomous/TimedSmartDuck.java (hunk 4 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/autonomous/TimedSmartDuck.java]: Got it

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/ColorSensorTelemetry.java (hunk 5 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/ColorSensorTelemetry.java]: I have not idea 🤷‍♂️

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/Control.java (hunk 6 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/Control.java]: Initially, it was going to be some sort of enum for using the Controller class and TeleOpAlt, but I never found a good way to use it. I will probably delete it.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpAlt.java (hunk 7 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpAlt.java]: Yeah, as we continue I think maintaining and adjusting TeleOpAlt is going to be hard. I've been debating whether to rewrite it, delete it, or not. Most likely I'll remove it and clean up TeleOpMain to be more readable.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpMain.java (hunk 8 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpMain.java]: I haven't seen an issue with this specific to the arm, there have only been issues with the claw. Is this something that needs to change?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpAlt.java (hunk 7 above)
Comment (9p4) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpAlt.java]: If you take a look at https://github.com/BotsBurgh/BOTSBURGH-FTC-2021-22/blob/main/FtcRobotController/src/main/java/org/firstinspires/ftc/robotcontroller/external/samples/BasicOpMode_Iterative.java, you can get some ideas on how you can implement something sort of event driven that won't have these issues. I've been meaning to switch but I never got around to it.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpMain.java (hunk 8 above)
Comment (9p4) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpMain.java]: The speed of the arm is dependent on what else the robot is doing in that loop. By switching to the new method, it should get rid of a lot of invariability.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java (hunk 1 above)
Comment (9p4) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:54]: below those lines sorry

**Review by BD103 — COMMENTED** [review]
👍 

## Edit
I tried to comment on the commit, but it did a review instead. ¯\\\_(ツ)\_/¯

**Review by 9p4 — APPROVED** [review]

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java (hunk 1 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:54]: Fixed in 6bd254e

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/hw/SmartServo.java (hunk 3 above)
Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/hw/SmartServo.java]: I'm unsure how to do this, you may have to
